From 0fc81c40b40aec0c8b70e7bd274685767d394cbf Mon Sep 17 00:00:00 2001
From: Isaac Davis <isaac.davis@joyent.com>
Date: Fri, 5 Jul 2019 21:01:32 +0000
Subject: [PATCH] MANTA-4381 Add "buckets" shard type to manta Reviewed by:
 Kelly McLaughlin <kelly.mclaughlin@joyent.com> Approved by: Kelly McLaughlin
 <kelly.mclaughlin@joyent.com>

---
 cmd/manta-shardadm.js | 41 +++++++++++++++++++++++++++++++----------
 lib/common.js         |  1 +
 2 files changed, 32 insertions(+), 10 deletions(-)

diff --git a/cmd/manta-shardadm.js b/cmd/manta-shardadm.js
index a3fb50b..52b77d4 100755
--- a/cmd/manta-shardadm.js
+++ b/cmd/manta-shardadm.js
@@ -124,7 +124,7 @@ function printShards(metadata, cb) {
 Shardadm.prototype.do_set = function (subcmd, opts, args, cb) {
 	var self = this;
 
-	if (args.length !== 0 || (!opts.i && !opts.m && !opts.s)) {
+	if (args.length !== 0 || (!opts.i && !opts.b && !opts.m && !opts.s)) {
 		this.do_help('help', {}, [subcmd], cb);
 		return;
 	}
@@ -157,16 +157,11 @@ Shardadm.prototype.do_set = function (subcmd, opts, args, cb) {
 		}
 
 		if (opts.i) {
-			var names = opts.i.split(' ');
-			var shards = [];
-
-			names.forEach(function (name) {
-				var shard = addSuffix(name, domain_name);
-				shards.push({ host: shard });
-			});
-			shards[shards.length - 1].last = true;
+			addIndexShards(opts.i, common.INDEX_SHARDS);
+		}
 
-			metadata[common.INDEX_SHARDS] = shards;
+		if (opts.b) {
+			addIndexShards(opts.b, common.BUCKETS_SHARDS);
 		}
 
 		if (Object.keys(metadata).length === 0) {
@@ -182,6 +177,28 @@ Shardadm.prototype.do_set = function (subcmd, opts, args, cb) {
 			console.log('Updated Manta shards successfully');
 			return (cb(null));
 		    });
+
+		/*
+		 * Helper function for adding index shards to metadata object.
+		 * Arguments are:
+		 *
+		 * - nameStr: a string of shard names separated by spaces
+		 *
+		 * - key: the field in the metadata object in which to store
+		 *   the shard array
+		 */
+		function addIndexShards(nameStr, key) {
+			var names = nameStr.split(' ');
+			var shards = [];
+
+			names.forEach(function (name) {
+				var shard = addSuffix(name, domain_name);
+				shards.push({ host: shard });
+			});
+			shards[shards.length - 1].last = true;
+
+			metadata[key] = shards;
+		}
 	});
 };
 Shardadm.prototype.do_set.options = [
@@ -189,6 +206,10 @@ Shardadm.prototype.do_set.options = [
 	    names: [ 'i' ],
 	    type: 'string',
 	    help: 'shards for indexing tier'
+	}, {
+	    names: [ 'b' ],
+	    type: 'string',
+	    help: 'shards for manta buckets subsystem indexing tier'
 	}, {
 	    names: [ 'm' ],
 	    type: 'string',
diff --git a/lib/common.js b/lib/common.js
index 4c92cc5..68634ab 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -42,6 +42,7 @@ exports.MARLIN_DIR = '/opt/smartdc/agents/lib/node_modules/marlin';
 exports.MARLIN_SHARD = 'MARLIN_MORAY_SHARD';
 exports.STORAGE_SHARD = 'STORAGE_MORAY_SHARD';
 exports.INDEX_SHARDS = 'INDEX_MORAY_SHARDS';
+exports.BUCKETS_SHARDS = 'BUCKETS_MORAY_SHARDS';
 exports.HASH_RING_IMAGE = 'HASH_RING_IMAGE';
 exports.HASH_RING_IMGAPI_SERVICE = 'HASH_RING_IMGAPI_SERVICE';
 
-- 
2.21.0

